File: Abgabe/LBAufgabe2/src/trendelenburg/de/liste/List.java
package trendelenburg.de.liste;

public class List {

	public boolean is_a_list(ListenElement liste) {
		// 1. & 2.
		return liste.is_a_list();
	}

	public ListenElement diffList(ListenElement liste1, ListenElement liste2) {
		ListenElement ergebnisListe = new ListenElement();
		return diffList_(liste1, liste2, ergebnisListe);
	}

	private ListenElement diffList_(ListenElement liste1, ListenElement liste2, ListenElement ergebnisListe) {
		if (liste1.getData() == null) {
			return ergebnisListe;
		}
		// 1.
		if (liste1.isNextEmpty()) {
			if (!liste2.contains(liste1.getData())) {
				ergebnisListe.add(liste1.getData());
			}
			return ergebnisListe;
		} else {
			// 2.
			if (liste2.contains(liste1.getData())) {
				// 2.a
				return diffList_(liste1.getNext(), liste2, ergebnisListe);
			} else {

				// 2.b
				ergebnisListe.add(liste1.getData());
				return diffList_(liste1.getNext(), liste2, ergebnisListe);
			}
		}
	}

	public boolean praefix(ListenElement liste1, ListenElement liste2) {
		// 1.
		if (liste1.isNextEmpty() || liste2.isNextEmpty()) {
			return false;
		}
		return praefix_(liste1, liste2);
	}

	private boolean praefix_(ListenElement liste1, ListenElement liste2) {
		if (liste1.isNextEmpty() && !liste2.isNextEmpty()) {
			// 2.b:i
			return true;
		} else if (liste2.isNextEmpty()) {
			return false;
		} else {
			// 2.
			if (liste1.getData().equals(liste2.getData())) {
				// 2.a
				return praefix_(liste1.getNext(), liste2.getNext());
			} else {
				// 2.b:ii
				return false;
			}
		}
	}

	public boolean suffix(ListenElement liste1, ListenElement liste2) {
		if (liste1.isNextEmpty() || liste2.isNextEmpty()) {
			return false;
		}
		// 1. & 2. & 3.
		return praefix(reverse(liste1), reverse(liste2));
	}

	public ListenElement reverse(ListenElement liste1) {
		ListenElement le = new ListenElement();
		return reverse_(liste1, le);
	}

	private ListenElement reverse_(ListenElement liste1, ListenElement returnList) {
		if (liste1.isNextEmpty()) {
			return returnList;
		} else {
			if (returnList.isNextEmpty()) {
				return reverse_(liste1.getNext(), new ListenElement(liste1.getData(), new ListenElement()));
			} else {
				return reverse_(liste1.getNext(), new ListenElement(liste1.getData(), returnList));
			}
		}
	}

	public boolean infix(ListenElement liste1, ListenElement liste2) {
		// 1.
		if (liste2.isNextEmpty()) {
			return false;
		}
		// 2.
		return infix_(liste1, liste2.getNext());
	}

	private boolean infix_(ListenElement liste1, ListenElement liste2) {

		// 3.
		if (liste2.isNextEmpty()) {
			return false;
		}
		// 4.
		if (praefix(liste1, liste2)) {
			return true;
			// 4.b.ii
		} else {
			return infix_(liste1, liste2.getNext());
		}
	}

	public int length(ListenElement liste) {
		int length = length_(liste, 0);
		return length;
	}

	private int length_(ListenElement liste, int length) {
		if (liste.isNextEmpty()) {
			return length;
		} else if (liste.isNextEmpty()) {
			length++;
			return length;
		} else {
			length++;
			return length_(liste.getNext(), length);
		}
	}

	public EoReturn eo_count(ListenElement liste1) {
		EoReturn returnElement = new EoReturn();
		eo_count_(liste1, returnElement);
		if (length(liste1) % 2 == 0) {
			returnElement.addEven();
		} else {
			returnElement.addOdd();
		}
		return returnElement;
	}

	private EoReturn eo_count_(ListenElement liste1, EoReturn eoR) {
		// 1.
		if (liste1.isNextEmpty()) {
			// 1.a
			return eoR;
		}
		// 1.b
		if (liste1.getData() instanceof ListenElement) {
			ListenElement subList = (ListenElement) liste1.getData();
			// 1.b.i
			if ((length(subList) % 2) == 0) {
				// 1.b.i.1
				eoR.addEven();
			} else {
				// 1.b.i.2
				eoR.addOdd();
			}
			// 1.b.i.3
			eo_count_(liste1.getNext(), eoR);
			// 1.b.i.4
			eo_count_(subList.getNext(), eoR);
			return eoR;

		} else {
			// 1.b.ii
			return eo_count_(liste1.getNext(), eoR);
		}

	}

	public ListenElement del_element(String position, ListenElement element, ListenElement liste) {

		// 1.
		if (position == "l") {
			return reverse(del_element_("e", element, reverse(liste), new ListenElement()));
		}
		return del_element_(position, element, liste, new ListenElement());
	}

	private ListenElement del_element_(String position, ListenElement element, ListenElement liste,
			ListenElement returnList) {
		// 5.
		if (liste.isNextEmpty()) {
			return returnList;
		} else {
			if (element.getNext().isNextEmpty()) {

				if (liste.getData().equals(element.getData())) {

					// 3.
					if (position == "a") {
						return del_element_(position, element, liste.getNext(), returnList);

						// 2.
					} else if (position == "e") {
						return del_element_("ende", element, liste.getNext(), returnList);

						// 4.
					} else {
						returnList.add(liste.getData());
						return del_element_(position, element, liste.getNext(), returnList);
					}
					// 4.
				} else {
					returnList.add(liste.getData());
					return del_element_(position, element, liste.getNext(), returnList);
				}
			} else {
				if (liste.getData().equals(element)) {

					// 3.
					if (position == "a") {
						return del_element_(position, element, liste.getNext(), returnList);

						// 2.
					} else if (position == "e") {
						return del_element_("ende", element, liste.getNext(), returnList);

						// 4.
					} else {
						returnList.add(liste.getData());
						return del_element_(position, element, liste.getNext(), returnList);
					}
					// 4.
				} else {
					returnList.add(liste.getData());
					return del_element_(position, element, liste.getNext(), returnList);
				}
			}
		}

	}

	public ListenElement substitute(String position, ListenElement element, ListenElement newElement,
			ListenElement liste) {

		// 1.
		if (position == "l") {
			// 1.a-b
			return reverse(substitute_("e", element, newElement, reverse(liste), new ListenElement()));
		}
		return substitute_(position, element, newElement, liste, new ListenElement());
	}

	private ListenElement substitute_(String position, ListenElement element, ListenElement newElement,
			ListenElement liste, ListenElement returnList) {
		// 5.
		if (liste.isNextEmpty()) {
			return returnList;
		} else {
			if (element.getNext().isNextEmpty()) {

				if (liste.getData().equals(element.getData())) {

					// 3.
					if (position == "a") {
						returnList.add(newElement.getData());
						return substitute_(position, element, newElement, liste.getNext(), returnList);

						// 2.
					} else if (position == "e") {
						returnList.add(newElement.getData());
						return substitute_("ende", element, newElement, liste.getNext(), returnList);

						// 4.
					} else {
						returnList.add(liste.getData());
						return substitute_(position, element, newElement, liste.getNext(), returnList);
					}
					// 4.
				} else {
					returnList.add(liste.getData());
					return substitute_(position, element, newElement, liste.getNext(), returnList);
				}
			} else {
				if (liste.getData().equals(element)) {

					// 3.
					if (position == "a") {
						returnList.add(newElement.getData());
						return substitute_(position, element, newElement, liste.getNext(), returnList);

						// 2.
					} else if (position == "e") {
						returnList.add(newElement.getData());
						return substitute_("ende", element, newElement, liste.getNext(), returnList);

						// 4.
					} else {
						returnList.add(liste.getData());
						return substitute_(position, element, newElement, liste.getNext(), returnList);
					}
					// 4.
				} else {
					returnList.add(liste.getData());
					return substitute_(position, element, newElement, liste.getNext(), returnList);
				}
			}
		}

	}

}

File: Abgabe/list.pl
% Autor: Ferdinand Trendelenburg
% Datum: 26.10.2017

% 1.
is_a_list([]).
% 2.
is_a_list([_H|T]) :-
    is_a_list(T).

diffList(L1, [], L1).
% 1.a
diffList([], _L2, []).
diffList(L1, L2, ErgList) :-
        difflist_add(L1, L2, [], ErgList).

difflist_add([], _L2, L3, L3).
difflist_add([H1|T1], L2, L3, ErgList) :-
        % 2.a
        member(H1, L2),
        difflist_add(T1, L2, L3, ErgList).
difflist_add([H1|T1], L2, L3, ErgList):-
        % 2.a
        \+member(H1, L2),
        append(L3, [H1], NewList),
        difflist_add(T1, L2, NewList, ErgList).



% 1.
praefix([H|T1], [H|T2]):-
        praefix_(T1, T2).

% 2. b.i
praefix_([], [_H|_T]).
% 2. a
praefix_([H|T1], [H|T2]):-
        praefix_(T1, T2).



suffix(L1, L2):-
        % 1.
        reverse(L1, NewL1),
        % 2.
        reverse(L2, NewL2),
        % 3.
        praefix(NewL1, NewL2).





% 1. & 2.
infix([H|T1], [_H2|T2]):-
        infix_([H|T1], T2).

                % 4.a
infix_([H|T], [H|T1]):-
        infix_1([H|T], [H|T1]).
% 4.b
infix_([H1|T], [_H|T1]):-
        infix_([H1|T], T1).

infix_1([], [_H|_T]).
infix_1([H|T1], [H|T2]):-
        infix_1(T1, T2).

        % -----------------------------------------------

del_element(Position, Element, List, _ReturnList):-
      del_element_(Position, Element, List, []).

% 2.
del_element_(e, Element, [Element|T], ReturnList):-
      del_element_(ende, Element, T, ReturnList).

% 3.
del_element_(a, Element, [Element|T], ReturnList):-
      del_element_(a, Element, T, ReturnList).

% 1.
del_element_(l, Element, List, ReturnList):-
      reverse(List, NewList),
      del_element_(e, Element, NewList, ReturnList).

% 5.
del_element_(_Position, _Element, [], _ReturnList).
% 4.
del_element_(Position, Element, [H|T], ReturnList):-
      del_element_(Position, Element, T, [H|ReturnList]).


substitute(Position, Element, NewElement, List, _ReturnList):-
      substitute_(Position, Element, NewElement, List, []).

% 2.
substitute_(e, Element, NewElement, [Element|T], ReturnList):-
      substitute_(ende, Element, NewElement, T, [NewElement|ReturnList]).

% 3.
substitute_(a, Element, NewElement, [Element|T], ReturnList):-
      substitute_(a, Element, NewElement, T, [NewElement|ReturnList]).

% 1.
substitute_(l, Element, NewElement, List, ReturnList):-
      reverse(List, NewList),
      substitute_(e, Element, NewElement, NewList, ReturnList).

% 5.
substitute_(_Position, _Element, _NewElement, [], _ReturnList).
% 4.
substitute_(Position, Element, NewElement, [H|T], ReturnList):-
      substitute_(Position, Element, NewElement, T, [H|ReturnList]).


% 1.a
eo_count(_,0,0).
% 1.b.i
eo_count(List,Even,Odd):-

  % 1.b.i
  is_list(List),
  length(List, Length),
  0 is Length mod 2,
  % 1.b.i.3
  eo_count_(List,Even1,Odd),
  % 1.b.i.1
  Even is Even1+1.


% 1.b.i
eo_count(List,Even,Odd):-

  % 1.b.i
  is_list(List),
  % 1.b.i.3
  eo_count_(List,Even,Odd1),
  % 1.b.i.2
  Odd is Odd1 + 1.

eo_count_([],0,0).
eo_count_([Head|Tail],Even,Odd):-
  % 1.b.i.3
  eo_count_(Tail,TailEven,TailOdd),
  % 1.b.i.4
  eo_count(Head,HeadEven,HeadOdd),
  Even is HeadEven + TailEven,
  Odd is HeadOdd + TailOdd.



